ncclient/operations/third_party/iosxe/rpc.py

```diff
@@ -1,9 +1,500 @@
+        if period and dampening_period:
+            raise YangPushError("Can only have one of period and dampening_period")
+        if (period and streamident) or (dampening_period and streamident):
+            raise YangPushError("Cannot combine custom stream with periodic or on-change")
+        if (period is None) and (dampening_period is None) and (streamident is None):
+            raise YangPushError("Must have at least one of period, dampening_period or streamident")
+        if streamident and not streamns:
+            raise YangPushError("Must specify namespace for custom stream")
+
+        #
+        # Try to construct request the "standard" way. However,
+        # doesn't work because cannot force lxml to include namespace
+        # values for identities.
+        #
+        # node = new_ele_ns("establish-subscription", IETF_EVENT_NOTIFICATIONS_NS)
+        # stream = sub_ele_ns(node, "stream", IETF_EVENT_NOTIFICATIONS_NS)
+        # stream.text = 'yp:yang-push'
+        # stream.attrib = { 'xmlns:yp': IETF_YANG_PUSH_NS }
+        # sub_ele_ns(node, "xpath-filter", IETF_YANG_PUSH_NS).text = xpath
+        # if period:
+        #     sub_ele_ns(node, "period", IETF_YANG_PUSH_NS).text = str(period)
+        # elif dampening_period:
+        #     sub_ele_ns(node, "dampening-period", IETF_YANG_PUSH_NS).text = str(dampening_period)
+
+        # Have to hack request as can't figure out how to force NS
+        # inclusion for identity values yet!
+        if streamident:
+            subst = {
+                'msgid': self._id,
+                'streamns': streamns,
+                'streamident': streamident,
+                'streamxpath': xpath,
+            }
+            rpc = stream_subscription_template.format(**subst)                
+        elif period:
+            substitutions = {
+                'msgid': self._id,
+                'period': period,
+                'streamxpath': xpath,
+            }
+            rpc = period_template.format(**substitutions)
+        else:
+            substitutions = {
+                'msgid': self._id,
+                'dampeningperiod': dampening_period,
+                'streamxpath': xpath,
+            }
+            rpc = dampening_period_template.format(**substitutions)
+
+        # Install the listener if necessary, error if there is a
+        # conflicting listener already installed, such as
+        # NotificationHandler. Note that any new event listener type
+        # will have to do something siimilar if they consume the same
+        # events.
+        if not hasattr(self.session, 'yang_push_listener'):
+            if self.session.has_event_listener:
+                raise AlreadyHasEventListener()
+            else:
+                self.session.yang_push_listener = YangPushListener()
+                self.session.add_listener(self.session.yang_push_listener)
+                self.session.has_event_listener = True
```

Comment:
Would be less bug-prone if `add_listener` and `remove_listener` managed the `has_event_listener` flag themselves, rather than requiring the caller to remember to set the flag appropriately after calling them.